### PR TITLE
Protect rebin

### PR DESCRIPTION
**Pull request by andLaing**
This PR reopens #658 with an added patch to protect
against attempts to rebin using only negative charge bins.

Tests are added to demonstrate the problems and their
solution.

### CODE REVIEW

File: invisible_cities/reco/peak_functions.py [hunk 1]
```diff
@@ -153,9 +153,10 @@ def rebin_times_and_waveforms(times, widths, waveforms,
     rebinned_wfs    = np.zeros((n_sensors, len(slices)))
 
     for i, s in enumerate(slices):
-        t  = times    [   s]
-        e  = waveforms[:, s]
-        w  = np.sum(e, axis=0) if np.any(e) else None
+        t = times    [   s]
+        e = waveforms[:, s]
+        w = np.sum(e, axis=0).clip(0) if np.any(e) else None
+        if not np.any(w): w = None
```
Comment (mmkekic) [inline invisible_cities/reco/peak_functions.py]: This is equivalent to a more readable version (at least to me):
```
w = np.sum(e, axis=0).clip(0) #note that we dont have to check if there are non-0s in e  since we will check the value of the sum later
if np.all(w == 0) : w = None
```
Maybe this can be made a bit more elegant so we dont rewrite `w`

**Review by mmkekic — COMMENTED** [review]
The 2 tests show ill-behaved old version of rebinning algorithm and the need to fix the bugs. The fix ensures that the bin position after rebinning will be physical. This PR still doesn't fix the problem of negative waveforms after the rebinning (this was intended in PR #683) , as I understand this is left to the future PR of @ausonandres  and @msorel.

File: invisible_cities/reco/peak_functions_test.py [hunk 2]
```diff
@@ -597,3 +600,12 @@ def test_rebin_times_and_waveforms_times_are_consistent(t_and_wf, stride):
                                                   np.ones_like(wfs), stride)
 
     assert np.sum(rb_times) * stride == approx(np.sum(times))
+
+
+def test_rebin_times_and_waveforms_negative_bins(ICDATADIR):
+    pmap_file = os.path.join(ICDATADIR, 'pmaps_negative_bins.h5')
```
Comment (mmkekic) [inline invisible_cities/reco/peak_functions_test.py:608]: maybe add a comment to explain what this test does? something like 'test that rebinning doesnt raise an error even when all the weights are negative". 
If in the future we introduce the masking of the negative signal we can reuse the test file to check if we indeed get rid of negative bins.

**Review by mmkekic — APPROVED** [review]
This PR demonstrates and  fixes the bug in weighted rebinning of the waveforms with negative weights. The change is minimal and clear, and the test are relevant for any kind of future refactoring of irene city. Good job!
